### PR TITLE
nutlink: init

### PR DESCRIPTION
**Pull request by mmahut**
Initial work for the `/nutlink` endpoint.

### CODE REVIEW

File: swagger.yaml [hunk 1]
```diff
@@ -3540,6 +3539,271 @@ paths:
         "500":
           $ref: "#/components/responses/internal_server_error"
 
+  /nutlink/{address}:
+    post:
```
Comment (mmahut) [inline swagger.yaml:3543]:
```suggestion
```

File: swagger.yaml [hunk 2]
```diff
@@ -3540,6 +3539,271 @@ paths:
...
+          $ref: "#/components/responses/not_found"
+        "429":
+          $ref: "#/components/responses/overusage_limit"
+        "418":
+          $ref: "#/components/responses/autobanned"
+        "500":
+          $ref: "#/components/responses/internal_server_error"
+
+  /nutlink/{address}/tickers:
+    post:
```
Comment (mmahut) [inline swagger.yaml:3575]:
```suggestion
```

File: swagger.yaml (hunk 1 above)
Comment (1000101) [inline swagger.yaml:3543]:
```suggestion
    get:
```

Comment (1000101) [inline swagger.yaml:3543]:
```suggestion
    get:
```

File: swagger.yaml (hunk 2 above)
Comment (1000101) [inline swagger.yaml:3575]:
```suggestion
    get:
```

File: swagger.yaml (hunk 1 above)
Comment (1000101) [inline swagger.yaml:3543]:
```suggestion
    get:
```

File: swagger.yaml [hunk 3]
```diff
@@ -3540,6 +3539,271 @@ paths:
...
+          $ref: "#/components/responses/not_found"
+        "429":
+          $ref: "#/components/responses/overusage_limit"
+        "418":
+          $ref: "#/components/responses/autobanned"
+        "500":
+          $ref: "#/components/responses/internal_server_error"
+
+  /nutlink/{address}/{ticker}:
+    post:
```
Comment (1000101) [inline swagger.yaml:3632]:
```suggestion
    get:
```

File: swagger.yaml [hunk 4]
```diff
@@ -3540,6 +3539,271 @@ paths:
...
+          $ref: "#/components/responses/not_found"
+        "429":
+          $ref: "#/components/responses/overusage_limit"
+        "418":
+          $ref: "#/components/responses/autobanned"
+        "500":
+          $ref: "#/components/responses/internal_server_error"
+
+  /nutlink/{ticker}/addresses:
+    post:
```
Comment (1000101) [inline swagger.yaml:3695]:
```suggestion
    get:
```

File: swagger.yaml [hunk 5]
```diff
@@ -3540,6 +3539,271 @@ paths:
...
+          $ref: "#/components/responses/not_found"
+        "429":
+          $ref: "#/components/responses/overusage_limit"
+        "418":
+          $ref: "#/components/responses/autobanned"
+        "500":
+          $ref: "#/components/responses/internal_server_error"
+
+  /nutlink/{ticker}:
+    post:
```
Comment (1000101) [inline swagger.yaml:3751]:
```suggestion
    get:
```

File: swagger.yaml [hunk 6]
```diff
@@ -5568,5 +5832,82 @@ components:
...
+    nutlink_address_tickers:
+      type: array
+      items:
+        type: object
+        properties:
+          name:
+            type: string
+            description: Name of the ticker
+          count:
+            type: integer
```
Comment (1000101) [inline swagger.yaml:5864]:
```suggestion
            type: string
```
Are you sure int won't overflow?

**Review by 1000101 — CHANGES_REQUESTED** [review]

File: swagger.yaml [hunk 7]
```diff
@@ -5568,5 +5832,82 @@ components:
...
+          description: URL do specific metadata file
+          example: "https://nut.link/metadata.json"
+        metadata_hash:
+          type: string
+          description: Hash of the metadata file
+          example: "6bf124f217d0e5a0a8adb1dbd8540e1334280d49ab861127868339f43b3948af"
+        metadata:
+          type: object
+          description: The cached metadata of the `metadata_url` file.
+          additionalProperties: true
```
Comment (1000101) [inline swagger.yaml:5849]:
```suggestion
          additionalProperties: true
```

do we need this if we don't define any specific requirements on the metadata object?

File: swagger.yaml [hunk 8]
```diff
@@ -5568,5 +5832,82 @@ components:
...
+          nullable: true
+      required:
+        - metadata_url
+        - metadata_hash
```
Comment (1000101) [inline swagger.yaml:5853]:
```suggestion
      required:
        - metadata_url
        - metadata_hash
        - metadata
```

I think metadata should always be present, despite the fact that they can be `null`

File: swagger.yaml [hunk 9]
```diff
@@ -5568,5 +5832,82 @@ components:
...
+            type: string
+            description: Block height of the latest record
+            example: "2657092"
+        required:
+          - address
+          - latest
+
+    nutlink_address_ticker:
+      type: object
+      additionalProperties: true
```
Comment (1000101) [inline swagger.yaml:5903]:
```suggestion
      additionalProperties: true
```
Same as with the other addProps - is this needed?

File: swagger.yaml [hunk 10]
```diff
@@ -5568,5 +5832,82 @@ components:
...
+        type: object
+        properties:
+          name:
+            type: string
+            description: Name of the ticker
+          count:
+            type: integer
+            description: Number of ticker records
+          latest:
+            type: string
```
Comment (1000101) [inline swagger.yaml:5867]:
```suggestion
            type: integer
```
Block height is integer.

File: swagger.yaml [hunk 11]
```diff
@@ -3540,6 +3539,271 @@ paths:
...
+        "404":
+          $ref: "#/components/responses/not_found"
+        "429":
+          $ref: "#/components/responses/overusage_limit"
+        "418":
+          $ref: "#/components/responses/autobanned"
+        "500":
+          $ref: "#/components/responses/internal_server_error"
+
+  /nutlink/{ticker}:
```
Comment (1000101) [inline swagger.yaml:3750]:
```suggestion
  /nutlink/tickers/{ticker}:
```
This route collides with `/nutlink/{address}`

File: swagger.yaml [hunk 12]
```diff
@@ -3540,6 +3539,271 @@ paths:
...
+        "404":
+          $ref: "#/components/responses/not_found"
+        "429":
+          $ref: "#/components/responses/overusage_limit"
+        "418":
+          $ref: "#/components/responses/autobanned"
+        "500":
+          $ref: "#/components/responses/internal_server_error"
+
+  /nutlink/{ticker}/addresses:
```
Comment (1000101) [inline swagger.yaml:3694]:
```suggestion
  /nutlink/tickers/{ticker}/addresses:
```

Might also add additional `tickers` as the outher route collides with `/nutlink/{address}`

**Review by 1000101 — CHANGES_REQUESTED** [review]

File: swagger.yaml (hunk 9 above)
Comment (1000101) [reply swagger.yaml:5903]: Ok, seems to be needed as without explicitly specifying the object structure, the schema won't validate.

File: swagger.yaml (hunk 7 above)
Comment (1000101) [reply swagger.yaml]: My bad, definitely needed - without specifying the object structure it won't validate. So we either need this or explicitly define what's in metadata - I would vouch for the latter, but I'm not sure if this format is a standard.

File: swagger.yaml [hunk 13]
```diff
@@ -5568,5 +5832,82 @@ components:
...
+          description: URL do specific metadata file
+          example: "https://nut.link/metadata.json"
+        metadata_hash:
+          type: string
+          description: Hash of the metadata file
+          example: "6bf124f217d0e5a0a8adb1dbd8540e1334280d49ab861127868339f43b3948af"
+        metadata:
+          type: object
+          description: The cached metadata of the `metadata_url` file.
+          nullable: true
```
Comment (1000101) [inline swagger.yaml:5849]:
```suggestion
          description: The cached metadata of the `metadata_url` file.
          additionalProperties: true
          nullable: true
```


My bad, definitely needed - without specifying the object structure it won't validate. So we either need this or explicitly define what's in metadata - I would vouch for the latter, but I'm not sure if this format is a standard.

File: swagger.yaml [hunk 14]
```diff
@@ -5568,5 +5832,83 @@ components:
...
+      type: array
+      items:
+        type: object
+        properties:
+          address:
+            type: string
+            description: Address of a metadata oracle
+            example: "addr_test1qpmtp5t0t5y6cqkaz7rfsyrx7mld77kpvksgkwm0p7en7qum7a589n30e80tclzrrnj8qr4qvzj6al0vpgtnmrkkksnqd8upj0"
+          latest:
+            type: string
```
Comment (1000101) [inline swagger.yaml:5895]:
```suggestion
          latest_block:
            type: integer
```

File: swagger.yaml [hunk 15]
```diff
@@ -5568,5 +5832,83 @@ components:
...
+            description: Block height of the latest record
+            example: "2657092"
+        required:
+          - address
+          - latest
```
Comment (1000101) [inline swagger.yaml:5900]:
```suggestion
          - latest_block
```

File: swagger.yaml [hunk 16]
```diff
@@ -5568,5 +5832,83 @@ components:
...
+      example:
+        - asset: "ADAUSD"
+          count: 1980038
+          latest: "2657092"
+        - asset: "ADAEUR"
+          count: 1980038
+          latest: "2657092"
+        - asset: "ADABTC"
+          count: 1980038
+          latest: "2657092"
```
Comment (1000101) [inline swagger.yaml:5883]:
```suggestion
          latest_block:
            type: integer
            description: Block height of the latest record
        required:
          - name
          - count
          - latest_block
      example:
        - asset: "ADAUSD"
          count: 1980038
          latest_block: 2657092
        - asset: "ADAEUR"
          count: 1980038
          latest_block: 2657092
        - asset: "ADABTC"
          count: 1980038
          latest_block: 2657092
```

**Review by 1000101 — CHANGES_REQUESTED** [review]

File: swagger.yaml [hunk 17]
```diff
@@ -5568,5 +5832,83 @@ components:
...
+          latest_block:
+            type: integer
+            description: Block height of the latest record
+            example: "2657092"
+        required:
+          - address
+          - latest
+
+    nutlink_address_ticker:
+      type: object
```
Comment (1000101) [inline swagger.yaml:5903]:
```suggestion
      type: array
```

**Pull request merged**